File: python_backend/api_helpers/get_turn_data.py
from game_logic.ranges.top.get_ranges import get_ranges
from game_logic.pathsInfo.top.get_pathdata_dict import get_pathdata_dict
from game_logic.getters.get_reset_piece_dicts import get_reset_piece_dicts
from game_logic.threatArea.get_king_locs import get_king_locs
from game_logic.threatArea.top.get_threat_area import get_threat_area
from game_logic.restriction.get_num_pieces_checking_king import get_num_pieces_checking_king
from game_logic.restriction.get_multithreat_restriction import get_multithreat_restriction
from game_logic.ranges.top.get_final_ranges import get_final_ranges
from game_logic.coordType.rankfile.map_xy_to_rf import map_xy_to_rf
from game_logic.pins.top.get_pins import get_pins
from game_logic.fenParser.GameStatus.status_update import status_update
from game_logic.color.get_next_color import get_next_color as get_enemy_color
from game_logic.fenParser.GameStatus.is_checkmate import is_checkmate as checkmate
from api_helpers.ai_move import ai_move


def get_turn_data(board, color, ai_color, json_records, piece_defs, id_dict):
    """data for player who's turn it is now, at current point in the game
    calculations:
    ............
    final ranges: where every piece of player's pieces can move to.
    status: is it check, checkmate, stalemate or none of these? is the game over?
    ai_start: the starting square of a move if it is the computer's turn
    ai_dest: the ending square of a move if it is the computer's turn
    ............
    """
    init_ranges, pins, mt_restricts, final_ranges = get_reset_piece_dicts(board, color)
    init_ranges, special_moves = get_ranges(board, color, init_ranges, json_records, piece_defs, id_dict)
    k_loc = get_king_locs(board, color)
    threat_area = get_threat_area(board, k_loc, color, piece_defs, id_dict)
    pd_dict = get_pathdata_dict(board, k_loc, color, piece_defs, id_dict)
    pins = get_pins(pd_dict, pins)
    npck = get_num_pieces_checking_king(board, k_loc, color, pd_dict, piece_defs, id_dict)
    mt_restricts = get_multithreat_restriction(board, npck, color)
    final_ranges = get_final_ranges(init_ranges, pins, threat_area, final_ranges, mt_restricts)
    special_moves.set_promos(board, final_ranges, color)

    if color == ai_color and not checkmate(final_ranges):
        ai_capture, ai_start, ai_dest = ai_move(board, final_ranges, ai_color, special_moves)
    else:
        ai_capture, ai_start, ai_dest = False, False, False
    data = {'status': status_update(board, final_ranges, get_enemy_color(color), npck)}
    data.update(map_xy_to_rf({"ranges": final_ranges,
                              "moves": special_moves.get_moves(),
                              "ai_start": ai_start,
                              "ai_dest": ai_dest,
                              "ai_capture": ai_capture}))
    return data

File: node.js/apiHelpers/getTurnData.js
const getRanges = require("../ranges/top/getRanges");
const getPathdataDict = require("../pathsInfo/top/getpathDataDict");
const getResetPieceDicts = require("../getters/getResetPieceDicts");
const getKingLocs = require("../threatArea/getKingLocs");
const getThreatArea = require("../threatArea/top/getThreatArea");
const getNumPiecesCheckingKing = require("../restriction/getNumPiecesCheckingKing");
const getMultithreatRestriction = require("../restriction/getMultithreatRestriction");
const getFinalRanges = require("../ranges/top/getFinalRanges");
const getPins = require("../pins/top/getPins");
const pieceDefs = require("../testObjects/standardPieceDefs");
const printBoard = require("../printers/printBoard");
// const mapXyToRf = require("../coordType/rankfile/mapXyToRf");
// const statusUpdate = require("../fenParser/GameStatus/statusUpdate");
// const aiMove = require("./aiMove");


function getTurnData(board, color, aiColor, jsonRecords, pieceDefs, idDict) {
    /**data for player who's turn it is now, at current the.includes(point) game
    calculations:
    ............
    final ranges: where every piece of player's pieces can move to.
    status: is it check, checkmate, stalemate or none of these? is the game over?
    aiStart: the starting square of a move if it is the computer's turn
    aiDest: the ending square of a move if it is the computer's turn
    ............
    */
    var [initRanges, pins, mtRestricts, finalRanges] = getResetPieceDicts(board, color);
    var [initRanges, specialMoves] = getRanges(board, color, initRanges, jsonRecords, pieceDefs, idDict);
    var kLoc = getKingLocs(board, color);
    var threatArea = getThreatArea(board, kLoc, color, pieceDefs, idDict);
    var pdDict = getPathdataDict(board, kLoc, color, pieceDefs, idDict);
    var pins = getPins(pdDict, pins);
    var npck = getNumPiecesCheckingKing(board, kLoc, color, pdDict, pieceDefs, idDict);
    var mtRestricts = getMultithreatRestriction(board, npck, color);
    finalRanges = getFinalRanges(initRanges, pins, threatArea, finalRanges, mtRestricts);
    specialMoves.setPromos(board, finalRanges, color);

    return [finalRanges, specialMoves]

    // var aiCapture, aiStart, aiDest;
    // if (color === aiColor && ! checkmate(finalRanges)) {
    //     var [aiCapture, aiStart, aiDest] = aiMove(board, finalRanges, aiColor, specialMoves);
    // }
    // else {
    //     var [aiCapture, aiStart, aiDest] = [false, false, false]
    // }

    // let data = {'status': statusUpdate(board, finalRanges, getEnemyColor(color), npck)}

    // data.update(mapXyToRf({   "ranges": finalRanges,
    //                           "moves": specialMoves.getMoves(),
    //                           "aiStart": aiStart,
    //                           "aiDest": aiDest,
    //                           "aiCapture": aiCapture}))
    // return data;
}

module.exports = getTurnData;

if (require.main === module) {

    const getStandardPieceDefs = require("../testObjects/getStandardPieceDefs")
    const getStandardIdDict = require("../testObjects/getStandardIdDict");
    const initEmptyRanges = require("./initEmptyRanges")
    const initDDEmptyLists = require("./initDDEmptyLists");
    const initPawnIds = require("../JsonRecords/initPawnIds");
    const sampleBoardDicts = require("../testObjects/sampleBoardDicts");
    const JsonRecords = require("../JsonRecords/JsonRecords");
    const getPieceLoc = require("../helpers/getPieceLoc")
    var dataDict = require("../testObjects/dataDicts/dd1");

    dataDict = initDDEmptyLists(dataDict)
    const pieceDefs = initEmptyRanges(getStandardPieceDefs())
    const idDict = getStandardIdDict()

    var board, jsonRecords;

    var game_names = [
      'castle_test1',
      'castle_test2',
      'castle_test3',
      'castle_test4',
      'check_example1',
      'check_example2',
      'check_example3',
      'check_example4',
      'check_example6',
      'dummy_game',
      'king_range_test',
      'knight_threat1',
      'knight_threat2',
      'pawn_promo',
      'pawn_range',
      'pawn_threat1',
      'ranges_test',
      'pinned_ex1',
      'pinned_ex2',
      'stalemate_exp1',
      'enpassant_test1',
      'fundemental_defense',
      'super_checkmate_impossible_example'
    ]

    for (var game_name of game_names) {
        console.log(`------------------------ ${game_name} ------------------------`.green)
        board = sampleBoardDicts[game_name]
        records = dataDict[game_name]["json"]
        jsonRecords = new JsonRecords(initPawnIds(records, board))
        var [finalRanges, specialMoves] = getTurnData(board, 'W', 'B', jsonRecords, pieceDefs, idDict)
        for (var id of Object.keys(finalRanges)) {
            printBoard(board, `\n${id}`, finalRanges[id], null, [getPieceLoc(board, id)])
        }
    }
}